vmdb/app/controllers/api_controller/providers.rb

```diff
@@ -0,0 +1,126 @@
+class ApiController
+  module Providers
+    CREDENTIAL_ATTRS = %w(userid password)
+    RESTRICTED_ATTRS = %w(type) + CREDENTIAL_ATTRS
+
+    def create_resource_providers(type, _id, data = {})
+      if data.key?("id") || data.key?("href")
+        raise BadRequestError,
```

Comment:
Ruby already has an exception for this...ArgumentError.  Does it make sense to raise proper errors in the code, and then if needed translate those to BadRequestError or perhaps some other HTTP error that matches?  http://stackoverflow.com/questions/9454811/which-http-status-code-to-use-for-required-parameters-not-provided